gravitee-am-service/src/main/java/io/gravitee/am/service/impl/PasswordHistoryService.java

```diff
@@ -0,0 +1,190 @@
+                                                                                 .user(user)
+                                                                                 .principal(principal)
+                                                                                 .type(EventType.PASSWORD_HISTORY_CREATED)
+                                                                                 .throwable(throwable))).toMaybe();
+
+    }
+
+    /**
+     * Checks if a password is already in the history.
+     *
+     * @param referenceType    the reference type (DOMAIN, APPLICATION)
+     * @param referenceId      id of the reference
+     * @param userId           id of user for this password
+     * @param password         the password to add
+     * @param passwordSettings domain/application password settings
+     * @return Single containing a {@link Boolean} {@code true} if the password is already present the allotted number of previous passwords.
+     */
+    public Single<Boolean> passwordAlreadyUsed(ReferenceType referenceType, String referenceId, String userId, String password, PasswordSettings passwordSettings) {
+        LOGGER.debug("Checking password history for user {}", userId);
+        if (passwordSettings == null || !passwordSettings.isPasswordHistoryEnabled()) {
+            return Single.just(false);
+        }
+        return repository.findUserHistory(referenceType, referenceId, userId)
+                         .toList()
+                         .flatMap(passwordHistories -> Single.just(passwordAlreadyUsed(password, passwordHistories)));
+    }
+
+    /**
+     * Returns a user's password history.
+     *
+     * @param referenceType    the reference type (DOMAIN, APPLICATION)
+     * @param referenceId      id of the reference
+     * @param userId           id of user for this password
+     * @return Flowable containing the user's password history.
+     */
+    public Flowable<PasswordHistory> findUserHistory(ReferenceType referenceType, String referenceId, String userId) {
+        return repository.findUserHistory(referenceType, referenceId, userId);
+    }
+
+    public Flowable<PasswordHistory> findByReference(ReferenceType referenceType, String referenceId) {
+        LOGGER.debug("Find password histories by reference id {} and reference type {}", referenceId, referenceType);
+        return repository.findByReference(referenceType, referenceId)
+                         .onErrorResumeNext(ex -> {
+                             LOGGER.error("Error finding password histories by reference id {} and reference type {}", referenceId, referenceType, ex);
+                             return Flowable.error(new TechnicalManagementException(
+                                     String.format("Error finding password histories by reference id %s and reference type %s", referenceId, referenceType), ex));
+                         });
+    }
+
+    public CompletableSource delete(String passwordHistoryId) {
```

Comment:
I would expect two method `deleteByUser(domainId, userId)` and `deleteByDomain(domainId)` to remove all the pwdHistory with a single request when a user is deleted or a domain is deleted.

In these cases, no need to trace an audit for each password history entries.

Reply:
I wanted to use the existing repository methods where possible to save time, rather than add new methods which will, in turn, require more tests to be written and generally take more time. So yes it would be more efficient to have a single repo call but to implement this is going to take considerably more time - ~ 1/2 day of work - to implement.

Reply:
it is OK to keep this method for the delete user part as for a user, we only have few entries to delete but for the delete domain, we have to develop this specific method to avoid too long operation and timeout.